src/api-service/__app__/onefuzzlib/pools.py

```diff
@@ -783,13 +783,16 @@ def cleanup_nodes(self) -> bool:
             logging.info("no nodes need updating: %s", self.scaleset_id)
             return False
 
+        removed = []
         # Nodes do not exists in scalesets but in table due to unknown failure
         for node in nodes:
             if node.machine_id not in azure_nodes:
+                removed.append(node.machine_id)
                 logging.info(
                     "no longer in scaleset: %s:%s", self.scaleset_id, node.machine_id
                 )
                 node.delete()
+        nodes = [x for x in nodes if x.machine_id not in nodes]
```

Comment:
Should the filter condition be `if x.machine_id not in removed`?

Suggestion: consider not shadowing/rebinding `nodes`, but using a fresh name that describes what is important about this list of nodes.

Reply:
Also, as written, it seems like the condition here should arguably be a type error, since `nodes` is something like a `List[Node]`, not `List[UUID]`.

Reply:
Good catch.  This should have checked `x.machine_id in azure_nodes`